pkg/registry/core/service/ipallocator/controller/repairip.go

```diff
@@ -523,6 +557,67 @@ func (r *RepairIPAddress) syncIPAddress(key string) error {
 	return nil
 
 }
+
+func (r *RepairIPAddress) cidrWorker() {
+	for r.processNextWorkCIDR() {
+	}
+}
+
+func (r *RepairIPAddress) processNextWorkCIDR() bool {
+	eKey, quit := r.cidrQueue.Get()
+	if quit {
+		return false
+	}
+	defer r.cidrQueue.Done(eKey)
+
+	err := r.syncCIDRs()
+	r.handleCIDRErr(err, eKey)
+
+	return true
+}
+
+func (r *RepairIPAddress) handleCIDRErr(err error, key interface{}) {
+	if err == nil {
+		r.cidrQueue.Forget(key)
+		return
+	}
+
+	if r.cidrQueue.NumRequeues(key) < maxRetries {
+		klog.V(2).InfoS("Error syncing ServiceCIDR, retrying", "serviceCIDR", key, "err", err)
+		r.cidrQueue.AddRateLimited(key)
+		return
+	}
+
+	klog.Warningf("Dropping ServiceCIDR %q out of the queue: %v", key, err)
+	r.cidrQueue.Forget(key)
+	runtime.HandleError(err)
+}
+
+// syncCIDRs rebuilds the radix tree based from the informers cache
+func (r *RepairIPAddress) syncCIDRs() error {
+	cidrList, err := r.serviceCIDRLister.List(labels.Everything())
+	if err != nil {
+		return err
+	}
+
+	treeV4 := iptree.New[string](false)
```

Comment:
the question remains :-) do we really need to rebuild the entire tree everytime.

Reply:
the problem is that we allow overlapping and duplicate ranges and we can get in an inconsistent state if we process deltas , if I create ServiceCIDR A with range 192.168.0.0/24 and ServiceCIDR B with the same range, I will have to take into account that there is still another one if one of the subnets is deleted.
This is not different than what we are doing with the bitmap allocators, we always rebuild them from scratch, and. is much simpler than taking into account all the existing ranges, also, we don't expect ServiceCIDRs go beyond the 2 digits, right?